xonsh/prompt/gitstatus.py

```diff
@@ -17,11 +17,27 @@
 
 def _check_output(*args, **kwargs):
     kwargs.update(dict(env=builtins.__xonsh_env__.detype(),
+                       stdout=subprocess.PIPE,
                        stderr=subprocess.DEVNULL,
-                       timeout=builtins.__xonsh_env__['VC_BRANCH_TIMEOUT'],
                        universal_newlines=True
                        ))
-    return subprocess.check_output(*args, **kwargs)
+    timeout = builtins.__xonsh_env__['VC_BRANCH_TIMEOUT']
+    # See https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate
+    with subprocess.Popen(*args, **kwargs) as proc:
+        try:
+            out, err = proc.communicate(timeout=timeout)
+            if proc.returncode != 0:
+                raise subprocess.CalledProcessError(proc.returncode, proc.args, output=out, stderr=err)  # note err will always be empty as we redirect stderr to DEVNULL abvoe
+            return out
+        except subprocess.TimeoutExpired:
+            # We use `.terminate()` (SIGTERM) instead of `.kill()` (SIGKILL) here
+            # because otherwise we guarantee that a `.git/index.lock` file will be
+            # left over, and subsequent git operations will fail.
+            # We don't want that.
+            # As a result, we must rely on git to exit properly on SIGTERM.
+            proc.terminate()
+            proc.wait()  # we ignore what git says after we sent it SIGTERM
```

Comment:
We need to wait, because otherwise `index.lock` will still exist, and so the next git operation (e.g. when you press Enter again, forcing xonsh to try generate a new gistatus) will fail with the `index.lock` error.

I believe that this is OK also from a user experience point of view, because git will exit quickly on SIGTERM. If it doesn't, it's a bug, and we have to know about it.

---

Technically, we don't even have to write `proc.wait()` here, because the `with Popen` context manager wil call `wait()` anyway when that block is left, but I wrote it anyway in order to be explicit about the waiting. I'll add a comment to that regard.